### ISSUE TITLE
Getting issue while running the sample code

### ISSUE THREAD

**Issue by rkaur83** [aws/aws-msk-iam-sasl-signer-net#14]
### Describe the issue

//Callback to handle OAuth bearer token refresh. It fetches the OAUTH Token from the AWSMSKAuthTokenGenerator class. 
   void OauthCallback(IClient client, string cfg)
   {
       try
       {
           var (token, expiryMs) = await mskAuthTokenGenerator.GenerateAuthTokenAsync(Amazon.RegionEndpoint.USEast1);
           client.OAuthBearerSetToken(token, expiryMs, "DummyPrincipal");
       }
       catch (Exception e)
       {
           client.OAuthBearerSetTokenFailure(e.ToString());
       }
   }

Above snippet needs to return a task, so it is giving error.

### Links

https://github.com/aws/aws-msk-iam-sasl-signer-net

**Comment by rkaur83** [comment]
Also, the handler in the below line require from the documentation requires Action as a parameter while the OauthCallback is a method group.

 var producer = new ProducerBuilder<string, string>(producerConfig)
                       .SetOAuthBearerTokenRefreshHandler(OauthCallback).Build();

**Comment by sankalpbhatia** [comment]
Thanks for reporting. I will check this in a bit. Meanwhile @rkaur83 do you have a proposed fix for this? or are you currently blocked?

**Comment by rkaur83** [comment]
@sankalpbhatia I have modified the code and successfully created a token using IAM authentication but getting "Time out" error while sending message via producer. Seems like there is either an issue with the network connectivity or generated token as I am using public endpoint of aws msk cluster but I am not sure. If you have any clue about it please let me know as there is nothing specific in the error to fix.

Fix for the code is given below.

 async void OauthCallback(IProducer<String, String> producer, string test=string.Empty)
            {
                try
                {
                    var (token, expiryMs) = await mskAuthTokenGenerator.GenerateAuthTokenFromRoleAsync(Amazon.RegionEndpoint.EUWest1, "arn:aws:iam::{account}:role/your-role-name");                 
                    }
                catch (Exception e)
                {
                    Console.WriteLine(e.ToString());
                }
            }
            _producer = new ProducerBuilder<string, string>(_producerConfig).SetErrorHandler(OnProduceError).SetOAuthBearerTokenRefreshHandler(oAuthBearerTokenRefreshHandler: (Action<IProducer<string,string>,string>)OauthCallback).Build();

**Comment by dulsi** [comment]
@rkaur83 An async function returns a Task not void. When the code triggers the OauthCallback function, it gets to the await and returns the Task object. The two variables are not set. If you then waited on the Task object it would get set. You don't seem to call client.OAuthBearerSetToken. Just getting the token isn't enough you need to tell the client.

This works in my case:
`void OauthCallback(IClient client, string cfg)
{
    try
    {
        Task<(string, long)> task = mskAuthTokenGenerator.GenerateAuthTokenAsync(Amazon.RegionEndpoint.USEast2);
        task.Wait();
        var (token, expiryMs) = task.Result;
        client.OAuthBearerSetToken(token, expiryMs, "DummyPrincipal");
    }
    catch (Exception e)
    {
        client.OAuthBearerSetTokenFailure(e.ToString());
    }
}`

**Comment by sankalpbhatia** [comment]
Thanks @dulsi. Another snippet which works is, by just adding the async keyword to the OauthCallback. I would argue that's a slightly better solution than what you shared as it conforms to C#/.NET best practices. 
```
           //Callback to handle OAuth bearer token refresh. It fetches the OAUTH Token from the AWSMSKAuthTokenGenerator class. 
            async void OauthCallback(IClient client, string cfg)
            {
                try
                {
                    var (token, expiryMs) = await mskAuthTokenGenerator.GenerateAuthTokenAsync(Amazon.RegionEndpoint.USEast1);
                    client.OAuthBearerSetToken(token, expiryMs, "DummyPrincipal");
                }
                catch (Exception e)
                {
                    client.OAuthBearerSetTokenFailure(e.ToString());
                }
            }

            var producer = new ProducerBuilder<string,string>(producerConfig).SetOAuthBearerTokenRefreshHandler(OauthCallback).Build();
            try
            {
                var deliveryReport = await producer.ProduceAsync("test-topic-msk-iam", new Message<string, string>
                {
                    Value = "Hello from .NET application"
                });

                Console.WriteLine($"Produced message to {deliveryReport.TopicPartitionOffset}");
            }
            catch (ProduceException<string, string> e)
            {
                Console.WriteLine($"failed to deliver message: {e.Message} [{e.Error.Code}]");
            }
```

@rkaur83 Again, thanks for reporting this miss. Can you try the snippet I shared above? I'll update documentation shortly.

**Comment by dulsi** [comment]
> Thanks @dulsi. Another snippet which works is, by just adding the async keyword to the OauthCallback. I would argue that's a slightly better solution than what you shared as it conforms to C#/.NET best practices.

What best practices?
https://learn.microsoft.com/en-us/archive/msdn-magazine/2013/march/async-await-best-practices-in-asynchronous-programming
"Prefer async Task methods over async void methods"

The problem is that somewhere the code is calling OauthCallback. The code assume after making that call the bearer token has been set. That assumption is wrong if it is async. You cannot guarantee that OauthCallback has completed. All you know is that it started. It may work perfectly fine in all your tests but it may fail randomly.

**Comment by robertcoltheart** [comment]
The main problem is that the underlying C# Kafka library does not provide an `async` version of the configuration setup for `SetOAuthBearerTokenRefreshHandler` (see [here](https://github.com/confluentinc/confluent-kafka-dotnet/blob/master/src/Confluent.Kafka/ProducerBuilder.cs#L284)). This means that you _must_ call `.Wait()` or `.Result` on the `GenerateAuthTokenAsync` method in order to get the token. 

In other words, either C# Kafka lib needs to provide an async callback for the callback handler, or this library needs to provide a sync version of `GenerateAuthTokenAsync`, eg. `GenerateAuthToken`.

**Comment by rkaur83** [comment]
Thanks @sankalpbhatia, this code snippet is working fine.

**Comment by sankalpbhatia** [comment]
Thanks @dulsi and @robertcoltheart for your inputs. We are discussing this internally and will share an update shortly.

**Comment by sankalpbhatia** [comment]
@dulsi @robertcoltheart We will be adding sync versions of GenerateAuthToken methods to this library in the next revision. Will share more updates soon.

**Comment by OrgesKreka** [comment]
@dulsi  I also noticed some difference in method signature from the one shown in the example.
If you need help i can write the mentioned change for sync versions.

**Comment by sankalpbhatia** [comment]
@OrgesKreka thanks for offering to help. Feel free to pick it up

**Comment by OrgesKreka** [comment]
> @OrgesKreka thanks for offering to help. Feel free to pick it up

@sankalpbhatia here is the pr: https://github.com/aws/aws-msk-iam-sasl-signer-net/pull/15

**Comment by sankalpbhatia** [comment]
update: The PR is merged and we working to release it next week.

**Comment by sankalpbhatia** [comment]
New version is released https://www.nuget.org/packages/AWS.MSK.Auth/1.0.1
